gazelle/resolve.go

```diff
@@ -0,0 +1,183 @@
+// rule generated by language.GenerateRules in
+// language.GenerateResult.Imports. Resolve generates a "deps" attribute (or
+// the appropriate language-specific equivalent) for each import according to
+// language-specific rules and heuristics.
+func (ts *TypeScript) Resolve(
+	c *config.Config,
+	ix *resolve.RuleIndex,
+	rc *repo.RemoteCache,
+	r *rule.Rule,
+	modulesRaw interface{},
+	from label.Label,
+) {
+	deps := ts.ResolveModuleDeps(c, ix, modulesRaw.(*treeset.Set), from)
+
+	DEBUG("RESOLVED(%s): %s => %s", from.Name, modulesRaw.(*treeset.Set).Values(), deps.Values())
+
+	if !deps.Empty() {
+		r.SetAttr("deps", convertDependencySetToExpr(deps))
+	}
+}
+
+func (ts *TypeScript) ResolveModuleDeps(
+	c *config.Config,
+	ix *resolve.RuleIndex,
+	modules *treeset.Set,
+	from label.Label,
+) *treeset.Set {
+	deps := treeset.NewWithStringComparator()
+
+	cfgs := c.Exts[languageName].(Configs)
+	cfg := cfgs[from.Pkg]
+	hasFatalError := false
+
+	it := modules.Iterator()
+	for it.Next() {
+		mod := it.Value().(ImportStatement)
+		imp := resolve.ImportSpec{
+			Lang: languageName,
+			Imp:  mod.Path,
+		}
+
+		DEBUG("RESOLVE: %q from %q", imp.Imp, from.Name)
+
+		if override, ok := resolve.FindRuleWithOverride(c, imp, languageName); ok {
+			if override.Repo == "" {
+				override.Repo = from.Repo
+			}
+			if !override.Equal(from) {
+				if override.Repo == from.Repo {
+					override.Repo = ""
+				}
+				dep := override.String()
+				deps.Add(dep)
+			}
+		} else if matches := ix.FindRulesByImportWithConfig(c, imp, languageName); len(matches) > 0 {
+			filteredMatches := make([]resolve.FindResult, 0, len(matches))
+			for _, match := range matches {
+				// Prevent from adding itself as a dependency.
+				if !match.IsSelfImport(from) {
+					filteredMatches = append(filteredMatches, match)
+				}
+			}
+
+			DEBUG("MATCHES(%q): %s", from.Name, filteredMatches)
+
+			if len(filteredMatches) == 1 {
+				matchLabel := filteredMatches[0].Label.Rel(from.Repo, from.Pkg)
+				dep := matchLabel.String()
+				deps.Add(dep)
+			} else if len(filteredMatches) > 1 {
+				err := fmt.Errorf(
+					"multiple targets (%s) may be imported with %q in %q "+
+						"- this must be fixed using the \"gazelle:resolve\" directive",
+					targetListFromResults(filteredMatches), mod.Path, mod.SourcePath)
+				log.Println("ERROR: ", err)
+				hasFatalError = true
+			}
+		} else if cfg.ValidateImportStatements() {
+			err := fmt.Errorf(
+				"Import %[1]q from %[2]q is an unknown dependency. Possible solutions:\n"+
+					"\t1. Instruct Gazelle to resolve to a known dependency using the gazelle:resolve directive.\n"+
+					"\t2. Ignore the dependency with a comment '# gazelle:%[3]s %[1]s' in the BUILD file",
+				mod.Path, mod.SourcePath, IgnoreImportsDirective,
+			)
+			log.Printf("ERROR: Failed to validate dependencies for target %q. %v", from.String(), err)
+			hasFatalError = true
+		}
+	}
+
+	if hasFatalError {
+		os.Exit(1)
```

Comment:
Same as the comment before re interrupting the program.

Reply:
Since we are just a gazelle-plugin I guess this means we are killing gazelle... do you know if there's a standard way to indicate an error to gazelle?

Reply:
I'm moving this up to the plugin entrypoint `Resolve()` method